### PR TITLE
FM2-290: Add Integration Tests for Person

### PR DESCRIPTION
**Pull request by gitcliff**
LINK:https://issues.openmrs.org/browse/FM2-290

### CODE REVIEW

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasProperty("fullUrl", startsWith("http://localhost/ws/fhir2/R3/Person/"))));
+		assertThat(entries, everyItem(hasResource(instanceOf(Person.class))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+	}
+	
+	@Test
+	public void shouldReturnSortedAndFilteredSearchResultsForPersonsAsXML() throws Exception {
+		MockHttpServletResponse response = get("/Person?family=Doe&_sort=given").accept(FhirMediaTypes.XML).go();
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:433]: You get an error for this test because "family" is not a configured endpoint for the Person resource. You might as well try: 
`get("/Person?name=Doe&_sort=given")`

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasProperty("fullUrl", startsWith("http://localhost/ws/fhir2/R3/Person/"))));
+		assertThat(entries, everyItem(hasResource(instanceOf(Person.class))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+	}
+	
+	@Test
+	public void shouldReturnSortedAndFilteredSearchResultsForPersonsAsJson() throws Exception {
+		MockHttpServletResponse response = get("/Person?family=Doe&_sort=given").accept(FhirMediaTypes.JSON).go();
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:388]: You get an error for this test because "family" is not a configured endpoint for the Person resource. You might as well try:
get("/Person?name=Doe&_sort=given")

File: test-data/src/test/resources/org/openmrs/module/fhir2/providers/PersonWebTest_create.xml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        <given value="Adam"/>
+    </name>
+    <gender value="male"/>
+    <birthDate value="2004-08-12"/>
+    <address>
+        <city value="Kampala"/>
+        <state value="Mukono"/>
+        <country value="Uganda"/>
+    </address>
+</Person>
```
Comment (varung-31) [inline test-data/src/test/resources/org/openmrs/module/fhir2/providers/PersonWebTest_create.xml:25]: Please leave a new line here

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PersonFhirResourceProviderIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package org.openmrs.module.fhir2.providers.r4;
+
+public class PersonFhirResourceProviderIntegrationTest {
+
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PersonFhirResourceProviderIntegrationTest.java:4]: Please complete this as well before marking the PR as ready for review. You might make it a draft for now

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+import org.hl7.fhir.dstu3.model.Person;
+import org.junit.Before;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+public class PersonFhirResourceProviderIntegrationTest extends BaseFhirR3IntegrationTest<PersonFhirResourceProvider, Person> {
+	
+	private static final String[] PERSON_SEARCH_DATA_FILES = {
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:49]: This can be written as a private static final variable since there is only one test data file to execute.

**Review by varung-31 — CHANGES_REQUESTED** [review]
@gitcliff Thanks for your work. I've added a few comments to get started.

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PersonFhirResourceProviderIntegrationTest.java (hunk 4 above)
Comment (gitcliff) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PersonFhirResourceProviderIntegrationTest.java:4]: yes @varung-31 am going to implement r4 to ,,wanted to first deal with errors in r3

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+		
+		assertThat(operationOutcome, notNullValue());
+		assertThat(operationOutcome.hasIssue(), is(true));
+	}
+	
+	@Test
+	public void shouldDeleteExistingPerson() throws Exception {
+		MockHttpServletResponse response = delete("/Person/" + PERSON_UUID).accept(FhirMediaTypes.JSON).go();
+		
+		assertThat(response, isOk());
```
Comment (gitcliff) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:344]: @varung-31 when i look in to the current error ie 
```

[ERROR]   PersonFhirResourceProviderIntegrationTest.shouldDeleteExistingPerson:344                                                                                                               Expected: response with HTTP status indicating request was handled successfully                                                                                                                       but: response with status code <410> with message "61b38324-e2fd-4feb-95b7-9e9a2a4400df"
--
```
throw on this method the uuid is deleted successfully, thats why we have  `code <410> with message "61b38324-e2fd-4feb-95b7-9e9a2a4400df"` tho am not quite sure as to why the response message here is negative 
--

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java (hunk 1 above)
Comment (ibacher) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:433]: Hmmm... that's actually a bug in the endpoint. We should support all the standard name properties on Person too.

Comment (varung-31) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:433]: @ibacher I don't think we can do that. That's because FHIR itself doesn't allow given and family for the person resource.
http://hl7.org/implement/standards/fhir/person.html

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java (hunk 6 above)
Comment (ibacher) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:344]: Admittedly, we should improve the message here somewhat, especially as I don't think that response should be returned from a `delete()` operation. So what happens in [`delete()`](https://github.com/openmrs/openmrs-module-fhir2/blob/6724bedc2730c379e6904fc7544916834014689e/api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseFhirDao.java#L125) is first we `get()` the result. In [`get()`](https://github.com/openmrs/openmrs-module-fhir2/blob/6724bedc2730c379e6904fc7544916834014689e/api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseFhirDao.java#L96), if the result is retired or voided already, we throw an exception. This is to ensure that an attempt to `get()` a record after it's been delete generates an error correctly.

The short version is that you're essentially asking to delete an already-voided Person record, so maybe try testing the functionality with a different Person. Nonetheless, that's a valuable error to have brought out.

Comment (varung-31) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:344]: @gitcliff i don't think that is the case. A 410 error means that the resource you are trying to delete is GONE. I think it might have been voided somewhere. We need to check where the bug is.

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+		Person person = readResponse(response);
+		
+		// update the existing record
+		Date birthDate = DateUtils.truncate(new Date(), Calendar.DATE);
+		person.setBirthDate(birthDate);
+		
+		// send the update to the server
+		response = put("/Person/" + PERSON_UUID).jsonContent(toJson(person)).go();
+		
+		assertThat(response, isOk());
```
Comment (gitcliff) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:223]: @varung-31  the same goes here with this [error](https://pastebin.com/k1HzWgNK)  of which i think the person must be some how connect to a particular patient in the system

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java (hunk 1 above)
Comment (ibacher) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PersonFhirResourceProviderIntegrationTest.java:433]: I guess that's why they aren't there!... Well that's a bit silly, but ok.

**Comment by ibacher** [comment]
@gitcliff Could you try rebasing your branch on top of the latest master? Hopefully my recent commit fixes some of these issues.

**Comment by gitcliff** [comment]
hello @ibacher,  sorry about the delayed response..
i did the rebase and it fixed the `shouldDeleteExistingPerson()` method of the resource however the 2 tests for update methods of xml and json data formats are still [failing ](https://pastebin.com/86n5ssrB) respectively

**Comment by varung-31** [comment]
> hello @ibacher, sorry about the delayed response..
> i did the rebase and it fixed the `shouldDeleteExistingPerson()` method of the resource however the 2 tests for update methods of xml and json data formats are still [failing ](https://pastebin.com/86n5ssrB) respectively.
> cc @varung-31

@gitcliff Could you please try update using a different Person? I think it's failing because person id = 2 has a patient associated with it that has void reason = "". It should work with other Persons (id = 3 or 4 or 5) in my opinion.

File: test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirPersonDaoImplTest_initial_data.xml [hunk 8]
```diff
@@ -9,8 +9,8 @@
     graphic logo is a trademark of OpenMRS Inc.
 -->
 <dataset>
-    <person person_id="2" gender="M" birthdate="1999-12-20" dead="false" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="61b38324-e2fd-4feb-95b7-9e9a2a4400df"/>
-    <person person_id="3" gender="M" birthdate="2000-11-18" dead="true" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="5c521595-4e12-46b0-8248-b8f2d3697766"/>
+    <person person_id="2" gender="M" birthdate="1999-12-20" dead="false" creator="1" date_created="2005-01-01 00:00:00.0" voided="false"  uuid="61b38324-e2fd-4feb-95b7-9e9a2a4400df"/>
+    <person person_id="3" gender="M" birthdate="2000-11-18" dead="false" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" void_reason="there is no reason" uuid="5c521595-4e12-46b0-8248-b8f2d3697766"/>
```
Comment (varung-31) [inline test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirPersonDaoImplTest_initial_data.xml:13]: Instead of changing this, please use the person with person_id=4

**Review by varung-31 — CHANGES_REQUESTED** [review]

File: test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirPersonDaoImplTest_initial_data.xml (hunk 8 above)
Comment (gitcliff) [reply test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirPersonDaoImplTest_initial_data.xml:13]: @varung-31 thanks for the review though i see no problem with making the change if it doesn't break any code or tests and it works well with the test

Comment (varung-31) [reply test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirPersonDaoImplTest_initial_data.xml:13]: @gitcliff Yes, but why change it if your code can work without this change ?

Comment (gitcliff) [reply test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirPersonDaoImplTest_initial_data.xml:13]: @varung-31  the `person_id=4` doesn't work to ,,thats why i used `person_id="3"`

**Review by ibacher — APPROVED** [review]
